exchange/src/exchange/bots/bot_container.cpp

```diff
@@ -72,33 +73,21 @@ BotContainer::add_retail_bot_(double starting_capital)
 {
     assert(starting_capital > 0);
     manager::ClientManager& users = nutc::manager::ClientManager::get_instance();
-    std::string bot_id = users.add_bot_trader();
-
-    retail_bots_.emplace(
-        std::piecewise_construct, std::forward_as_tuple(bot_id),
-        std::forward_as_tuple(bot_id, starting_capital)
-    );
-    const auto& bot = users.get_trader(bot_id);
-    assert(bot->get_type() == manager::BOT);
-    bot->set_capital(starting_capital);
-    // bot->modify_holdings(ticker_, INFINITY);
+    RetailBot bot(ticker_, starting_capital);
+    std::string bot_id = bot.get_id();
+    auto retail_bot = users.add_bot_trader(std::move(bot));
+    retail_bots_.insert({bot_id, retail_bot});
```

Comment:
i really don't like this API, and taking things as `T&&`'s when you are not doing c++ garbage is also a code smell

should be something like

```cpp
auto bot = users.add_retail_bot_trader(...);
retail_bots_.insert({bot.get_id(), std::move(bot)});
```

---

you can impl as follows

```cpp
template<class... Args>
std::shared_ptr<RetailBot>
add_retail_bot_trader(Args&&... args)
{
    auto bot = std::make_shared<RetailBot>(std::forward<Args>(args)...);
    add_bot_trader(bot);
    return bot;
}
```